shedskin/cpp.py

```diff
@@ -852,114 +869,130 @@ def instance_new(self, node, argtypes):
         ts_key = typestr(self.gx, self.subtypes(argtypes, 'unit'), mv=self.mv)
         ts_value = typestr(self.gx, self.subtypes(argtypes, 'value'), mv=self.mv)
-        for (key, value) in node.items:
+        items = list(zip(node.keys, node.values))
+        for (key, value) in items:
             self.visitm('(new tuple2<%s, %s>(2,' % (ts_key, ts_value), func)
             type_child = self.subtypes(argtypes, 'unit')
-            self.visit_conv(key, type_child, func)
+            self.impl_visit_conv(key, type_child, func)
             self.append(',')
             type_child = self.subtypes(argtypes, 'value')
-            self.visit_conv(value, type_child, func)
+            self.impl_visit_conv(value, type_child, func)
             self.append('))')
-            if (key, value) != node.items[-1]:
+            if (key, value) != items[-1]:
                 self.append(',')
         self.append('))')
 
     def visit_tuple_list(self, node, func=None, argtypes=None):
         if isinstance(func, Class):  # XXX
             func = None
         argtypes = self.instance_new(node, argtypes)
-        children = node.getChildNodes()
+        children = list(node.elts)
         if children:
             self.append(str(len(children)) + ',')
         if len(children) >= 2 and self.bin_tuple(argtypes):  # XXX >=2?
             type_child = self.subtypes(argtypes, 'first')
-            self.visit_conv(children[0], type_child, func)
+            self.impl_visit_conv(children[0], type_child, func)
             self.append(',')
             type_child = self.subtypes(argtypes, 'second')
-            self.visit_conv(children[1], type_child, func)
+            self.impl_visit_conv(children[1], type_child, func)
         else:
             for child in children:
                 type_child = self.subtypes(argtypes, 'unit')
-                self.visit_conv(child, type_child, func)
+                self.impl_visit_conv(child, type_child, func)
                 if child != children[-1]:
                     self.append(',')
         self.append('))')
 
-    def visitTuple(self, node, func=None, argtypes=None):
-        if len(node.nodes) > 2:
-            types = set()
-            for child in node.nodes:
-                types.update(self.mergeinh[child])
-            typestr(self.gx, types, node=child, tuple_check=True, mv=self.mv)
-        self.visit_tuple_list(node, func, argtypes)
-
-    def visitList(self, node, func=None, argtypes=None):
-        self.visit_tuple_list(node, func, argtypes)
+    def visit_Tuple(self, node, func=None, argtypes=None):
+        if type(node.ctx) == Load:
+            if len(node.elts) > 2:
+                types = set()
+                for child in node.elts:
+                    types.update(self.mergeinh[child])
+                typestr(self.gx, types, node=child, tuple_check=True, mv=self.mv)
+            self.visit_tuple_list(node, func, argtypes)
+        elif type(node.ctx) == Store:
+            assert False
```

Comment:
Does this mean that `node.ctx` can only be `Load`? Or are the other options just not implemented yet?
Could you add a comment about that to the code?

Reply:
> Does this mean that node.ctx can only be Load?

node.ctx can in general be `Load()`, `Store()`, and `Del()`.

@DevJohan, I love `assert False` myself in the in-development code (or even `assert 0`, or even `1/0`), but here `raise NotImplementedError` is a clear (self-documenting) winner.

Also, Python-idiomatic way to test being an instance of class is `isinstance(node.ctx, Store)`. (I'm not sure if shedskin might have another convention, and if you just follow the way of the original code; if not, I'd suggest to change.)

Reply:
That's a very naked `assert False`, sorry.

The reason there's an `assert False` there is Tuple with ctx=Store() is handled through the implementation of `visit_Assign` and this point is never reached. I don't think raising a `NotImplementedError` would be the correct message. If you want the same functionality as before, `generic_visit` should be called because previously `visitAssTuple`was not defined and would fall through to default behavior and visit every child node. I would rather just add a message saying "Implemented through visit_Assign, should not be reached".

When it comes to testing for ctx type I opted for `type(node.ctx) == XXX` because it's basically a type based enum. I want to test for a specific value not for instances of subclasses. According to the docs `expr_context = Load | Store | Del | AugLoad | AugStore | Param`, all are derived from the class `expr_context` but there's no internal class hierarchy if you query it.

Reply:
> I would rather just add a message saying "Implemented through visit_Assign, should not be reached".

+1, i.e. `assert False, "Implemented through visit_Assign, should not be reached"` (or look around the codebase to see if shedskin has a dedicated exception for such cases...)